code/modules/client/preferences/preferences.dm

```diff
@@ -0,0 +1,602 @@
+#define CHARACTER_SETUP 0
+#define UI_SETUP 1
+#define GENERAL_SETUP 2
+#define SPECIAL_ROLES_SETUP 3
+
+var/list/preferences_datums = list()
+var/global/list/special_roles = list(
+	ROLE_ALIEN     	= 1,
+	BLOBOVERMIND   	= 1,
+	ROLE_BORER     	= 1,
+	CHANGELING   	= 1,
+	CULTIST      	= 1,
+	ROLE_PLANT     	= 1,
+	MALF         	= 1,
+	NUKE_OP	    	= 1,
+	ROLE_PAI        = 1,
+	ROLE_POSIBRAIN  = 1,
+	REV          	= 1,
+	TRAITOR      	= 1,
+	CHALLENGER		= 1,
+	VAMPIRE      	= 1,
+	VOXRAIDER    	= 1,
+	WIZARD       	= 1,
+	ROLE_STRIKE	  	= 1,
+	GRINCH			= 1,
+	NINJA			= 1,
+	TIMEAGENT		= 1,
+	PULSEDEMON		= 1,
+	ROLE_MINOR		= 1,
+	ROLE_PRISONER   = 1,
+	ROLE_GRUE		= 1,
+	DIVERGENTCLONE  = 1,
+)
+
+/var/list/antag_roles = list(
+	ROLE_ALIEN      = 1,
+	BLOBOVERMIND   	= 1,
+	CHANGELING   	= 1,
+	CULTIST      	= 1,
+	MALF         	= 1,
+	NUKE_OP	    	= 1,
+	REV          	= 1,
+	TRAITOR      	= 1,
+	CHALLENGER		= 1,
+	VAMPIRE      	= 1,
+	VOXRAIDER    	= 1,
+	WIZARD       	= 1,
+	ROLE_STRIKE	  	= 1,
+	GRINCH			= 1,
+	NINJA			= 1,
+	TIMEAGENT		= 1,
+	PULSEDEMON		= 1,
+	ROLE_MINOR		= 1,
+	ROLE_PRISONER	= 1,
+	ROLE_GRUE		= 1,
+	DIVERGENTCLONE  = 1,
+)
+
+var/list/nonantag_roles = list(
+	ROLE_BORER        = 1,
+	ROLE_PLANT        = 1,
+	ROLE_PAI          = 1,
+	ROLE_POSIBRAIN    = 1,
+)
+
+var/list/role_wiki=list(
+	ROLE_ALIEN				= "Xenomorph",
+	BLOBOVERMIND			= "Blob",
+	ROLE_BORER				= "Cortical_Borer",
+	CHANGELING				= "Changeling",
+	CULTIST					= "Cult",
+	ROLE_PLANT				= "Dionaea",
+	MALF					= "Guide_to_Malfunction",
+	NUKE_OP					= "Nuclear_Agent",
+	ROLE_PAI				= "Personal_AI",
+	ROLE_POSIBRAIN			= "Guide_to_Silicon_Laws",
+	REV						= "Revolution",
+	TRAITOR					= "Traitor",
+	CHALLENGER				= "Challengers",
+	VAMPIRE					= "Vampire",
+	VOXRAIDER				= "Vox_Raider",
+	WIZARD					= "Wizard",
+	GRINCH					= "Grinch",
+	NINJA					= "Space_Ninja",
+	TIMEAGENT				= "Time_Agent",
+	PULSEDEMON				= "Pulse_Demon",
+	ROLE_MINOR				= "Minor_Roles",
+	ROLE_PRISONER			= "Minor_Roles",
+	ROLE_GRUE				= "Grue",
+	DIVERGENTCLONE			= "Divergent_Clone",
+)
+
+var/list/special_popup_text2num = list(
+	"Only use chat" = SPECIAL_POPUP_DISABLED,
+	"Only use special" = SPECIAL_POPUP_EXCLUSIVE,
+	"Use both chat and special" = SPECIAL_POPUP_USE_BOTH,
+)
+
+var/list/headset_sound_text2num = list(
+	"Disabled" = HEADSET_SOUND_DISABLED,
+	"Transmit Only" = HEADSET_SOUND_TRANSMIT,
+	"All" = HEADSET_SOUND_ALL,
+)
+
+var/const/MAX_SAVE_SLOTS = 16
+
+#define POLLED_LIMIT	100
+
+/datum/preferences
+	var/list/subsections
+	//doohickeys for savefiles
+	var/database/db = ("players2.sqlite")
+	var/path
+
+	// Which character slot
+	var/slot = 1
+	var/list/slot_names = new
+	var/slot_name = ""
+
+	// 0 = character settings, 1 = game preferences
+	var/current_tab = 0
+
+	// Last time the guy saved their prefs
+	var/lastPolled = 0
+
+	var/savefile_version = 0
+
+	// Alist = associative lists. This is a new 516 thing. Woo!
+	var/alist/preference_settings_client = alist()
```

Comment:
New BYOND features that works like a dict. Dreamchecker is going absolutely insane fighting these.
I can make it pass the linter but I'm losing the option of doing 
```dm
for (k, v in X)
```
which is basically the main draw (other than performance) of this new subtype.